includes/event/event-capabilities.php

```diff
@@ -0,0 +1,186 @@
+	private const EDIT   = 'edit_translation_event';
+	private const DELETE = 'delete_translation_event';
+
+	/**
+	 * All the capabilities that concern an Event.
+	 */
+	private const CAPS = array(
+		self::CREATE,
+		self::VIEW,
+		self::EDIT,
+		self::DELETE,
+	);
+
+	private Event_Repository_Interface $event_repository;
+	private Attendee_Repository $attendee_repository;
+	private Stats_Calculator $stats_calculator;
+
+	public function __construct(
+		Event_Repository_Interface $event_repository,
+		Attendee_Repository $attendee_repository,
+		Stats_Calculator $stats_calculator
+	) {
+		$this->event_repository    = $event_repository;
+		$this->attendee_repository = $attendee_repository;
+		$this->stats_calculator    = $stats_calculator;
+	}
+
+	/**
+	 * This function is automatically called whenever user_can() is called for one the capabilities in self::CAPS.
+	 *
+	 * @param string  $cap  Requested capability.
+	 * @param array   $args Arguments that accompany the requested capability check.
+	 * @param WP_User $user User for which we're evaluating the capability.
+	 * @return bool
+	 */
+	private function has_cap( string $cap, array $args, WP_User $user ): bool {
+		switch ( $cap ) {
+			case self::CREATE:
+				return $this->has_create( $user );
+			case self::VIEW:
+			case self::EDIT:
+			case self::DELETE:
+				if ( ! isset( $args[2] ) || ! is_int( $args[2] ) ) {
+					return false;
+				}
+				$event = $this->event_repository->get_event( $args[2] );
+				if ( ! $event ) {
+					return false;
+				}
+
+				if ( self::VIEW === $cap ) {
+					return $this->has_view( $user, $event );
+				} elseif ( self::EDIT === $cap ) {
+					return $this->has_edit( $user, $event );
+				} elseif ( self::DELETE === $cap ) {
+					return $this->has_delete( $user, $event );
+				}
+				break;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Evaluate whether a user can create events.
+	 *
+	 * @param WP_User $user User for which we're evaluating the capability.
+	 * @return bool
+	 */
+	private function has_create( WP_User $user ): bool {
+		return $this->is_gp_admin( $user );
+	}
+
+	/**
+	 * Evaluate whether a user can view a specific event.
+	 *
+	 * @param WP_User $user  User for which we're evaluating the capability.
+	 * @param Event   $event Event for which we're evaluating the capability.
+	 * @return bool
+	 */
+	private function has_view( WP_User $user, Event $event ): bool {
+		if ( $this->is_gp_admin( $user ) ) {
+			return true;
+		}
+
+		return 'publish' === $event->status();
+	}
+
+	/**
+	 * Evaluate whether a user can edit a specific event.
+	 *
+	 * @param WP_User $user  User for which we're evaluating the capability.
+	 * @param Event   $event Event for which we're evaluating the capability.
+	 * @return bool
+	 */
+	private function has_edit( WP_User $user, Event $event ): bool {
+		if ( $event->end()->is_in_the_past() ) {
+			return false;
+		}
+
+		if ( $this->stats_calculator->event_has_stats( $event->id() ) ) {
+			return false;
+		}
```

Comment:
Maybe this is too early and/or we need more fine grained controls (which is easy now!) But I think it should be possible to change at least the event description (and maybe the event end, if it should be extended) even when there are stats.

Reply:
Indeed, more fine grained capabilities could make sense, e.g. `edit_translation_event_with_stats`. Also agree that it it should be possible to edit certain fields of a past event. I opened https://github.com/WordPress/wporg-gp-translation-events/issues/227 to track that.